### ISSUE TITLE
`aws_api_gateway_stage` and `aws_api_gateway_method_settings` are not consistently handled

### ISSUE THREAD

**Issue by mikaelrandy** [hashicorp/terraform-provider-aws#5765]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version

```
$ terraform -v
Terraform v0.11.8
```

Provider version : `1.34.0`

### Affected Resource(s)

* aws_api_gateway_deployment
* aws_api_gateway_stage
* aws_api_gateway_method_settings

### Global explaination

I want to handle some API Gateway configurations with Terraform.
Theses configurations are available through `aws_api_gateway_stage` and `aws_api_gateway_method_settings` resources.

But there are 2 issues using theses resources : 
* `aws_api_gateway_stage` cannot be create in a whole API Gateway Terraform configuration (need to be imported)
* `aws_api_gateway_method_settings` and `aws_api_gateway_stage` configuration are dropped at each `aws_api_gateway_deployment` application.

### Terraform Configuration Files

```hcl
provider "aws" {
  region  = "eu-west-1"
  profile = "6cloud-dev"
  version = "~> 1.33"
}

resource "aws_api_gateway_rest_api" "test" {
  name        = "api_name"
  description = "Some API"
}

resource "aws_api_gateway_deployment" "test" {
  depends_on = [
    "aws_api_gateway_method.test_get",
  ]

  rest_api_id = "${aws_api_gateway_rest_api.test.id}"
  stage_name  = "test"
}

resource "aws_api_gateway_stage" "time" {
  rest_api_id   = "${aws_api_gateway_rest_api.test.id}"
  deployment_id = "${aws_api_gateway_deployment.test.id}"

  stage_name = "${aws_api_gateway_deployment.test.stage_name}"

  tags {
    project = "ProjectName"
  }
}

resource "aws_api_gateway_method_settings" "test" {
  rest_api_id = "${aws_api_gateway_rest_api.test.id}"
  stage_name  = "${aws_api_gateway_deployment.test.stage_name}"
  method_path = "*/*"

  settings {
    throttling_rate_limit = 1000
  }
}

resource "aws_api_gateway_resource" "test" {
  rest_api_id = "${aws_api_gateway_rest_api.test.id}"
  parent_id   = "${aws_api_gateway_rest_api.test.root_resource_id}"
  path_part   = "time"
}

resource "aws_api_gateway_method" "test_get" {
  rest_api_id   = "${aws_api_gateway_rest_api.test.id}"
  resource_id   = "${aws_api_gateway_resource.test.id}"
  http_method   = "GET"
  authorization = "NONE"
}

resource "aws_api_gateway_integration" "test" {
  rest_api_id = "${aws_api_gateway_rest_api.test.id}"
  resource_id = "${aws_api_gateway_resource.test.id}"
  http_method = "${aws_api_gateway_method.test_get.http_method}"
  type        = "MOCK"
}
```

### Debug Output

Complete gist : https://gist.github.com/mikaelrandy/a8c13555c7790742a03d865165d9f93a

* First `terraform apply` execution : [terraform_apigateway_stage_creation-tf](https://gist.github.com/mikaelrandy/a8c13555c7790742a03d865165d9f93a#file-terraform_apigateway_stage_creation-tf)
* Second `terraform apply`, after importing "aws_api_gateway_stage.time" : [terraform_apigateway_stage_update-tf](https://gist.github.com/mikaelrandy/a8c13555c7790742a03d865165d9f93a#file-terraform_apigateway_stage_update-tf)
* Third `terraform apply`, after taint "aws_api_gateway_deployment.test" : [terraform_apigateway_deployment_update-tf](https://gist.github.com/mikaelrandy/a8c13555c7790742a03d865165d9f93a#file-terraform_apigateway_deployment_update-tf)
* Last `terrafor apply`, to redeploy "aws_api_gateway_stage.time" : [terraform_apigateway_deployment_update_stage-tf](https://gist.github.com/mikaelrandy/a8c13555c7790742a03d865165d9f93a#file-terraform_apigateway_deployment_update_stage-tf)

### Expected Behavior

* At first deploy, all resources are deployed in AWS and correctly linked in tfstate
* At update, no configuration are lost

#### At first deploy
Execution fail on stage deployment. Seems that API Gateway stage is already created by another resource, but not linked.
```
terraform apply -auto-approve
aws_api_gateway_rest_api.test: Creating...
  api_key_source:           "" => "HEADER"
  created_date:             "" => "<computed>"
  description:              "" => "Some API"
  endpoint_configuration.#: "" => "<computed>"
  execution_arn:            "" => "<computed>"
  minimum_compression_size: "" => "-1"
  name:                     "" => "api_name"
  root_resource_id:         "" => "<computed>"
aws_api_gateway_rest_api.test: Creation complete after 0s (ID: wosvgb26dh)
aws_api_gateway_resource.test: Creating...
  parent_id:   "" => "a2f0vzet6l"
  path:        "" => "<computed>"
  path_part:   "" => "test"
  rest_api_id: "" => "wosvgb26dh"
aws_api_gateway_resource.test: Creation complete after 1s (ID: 8u42yk)
aws_api_gateway_method.test_get: Creating...
  api_key_required: "" => "false"
  authorization:    "" => "NONE"
  http_method:      "" => "GET"
  resource_id:      "" => "8u42yk"
  rest_api_id:      "" => "wosvgb26dh"
aws_api_gateway_method.test_get: Creation complete after 0s (ID: agm-wosvgb26dh-8u42yk-GET)
aws_api_gateway_deployment.test: Creating...
  created_date:  "" => "<computed>"
  execution_arn: "" => "<computed>"
  invoke_url:    "" => "<computed>"
  rest_api_id:   "" => "wosvgb26dh"
  stage_name:    "" => "test"
aws_api_gateway_integration.test: Creating...
  cache_namespace:      "" => "<computed>"
  connection_type:      "" => "INTERNET"
  http_method:          "" => "GET"
  passthrough_behavior: "" => "<computed>"
  resource_id:          "" => "8u42yk"
  rest_api_id:          "" => "wosvgb26dh"
  timeout_milliseconds: "" => "29000"
  type:                 "" => "MOCK"
aws_api_gateway_integration.test: Creation complete after 0s (ID: agi-wosvgb26dh-8u42yk-GET)
aws_api_gateway_deployment.test: Creation complete after 0s (ID: qhgl0a)
aws_api_gateway_stage.time: Creating...
  deployment_id: "" => "qhgl0a"
  execution_arn: "" => "<computed>"
  invoke_url:    "" => "<computed>"
  rest_api_id:   "" => "wosvgb26dh"
  stage_name:    "" => "test"
  tags.%:        "" => "1"
  tags.project:  "" => "ProjectName"
aws_api_gateway_method_settings.test: Creating...
  method_path:                      "" => "*/*"
  rest_api_id:                      "" => "wosvgb26dh"
  settings.#:                       "" => "1"
  settings.0.throttling_rate_limit: "" => "1000"
  stage_name:                       "" => "test"
aws_api_gateway_method_settings.test: Creation complete after 0s (ID: wosvgb26dh-test-*/*)

Error: Error applying plan:

1 error(s) occurred:

* aws_api_gateway_stage.time: 1 error(s) occurred:

* aws_api_gateway_stage.time: Error creating API Gateway Stage: ConflictException: Stage already exists
	status code: 409, request id: a9c1d588-b043-11e8-9a76-5fb0898636e5

Terraform does not automatically rollback in the face of errors.
Instead, your Terraform state file has been partially updated with
any resources that successfully completed. Please address the error
above and apply again to incrementally change your infrastructure.
```
Need to import stage and re-apply to succeed.

```
$ terraform import aws_api_gateway_stage.time wosvgb26dh/test
aws_api_gateway_stage.time: Importing from ID "wosvgb26dh/test"...
aws_api_gateway_stage.time: Import complete!
  Imported aws_api_gateway_stage (ID: ags-wosvgb26dh-test)
aws_api_gateway_stage.time: Refreshing state... (ID: ags-wosvgb26dh-test)

Import successful!

The resources that were imported are shown above. These resources are now in
your Terraform state and will henceforth be managed by Terraform.

$ terraform apply -auto-approve
aws_api_gateway_rest_api.test: Refreshing state... (ID: wosvgb26dh)
aws_api_gateway_resource.test: Refreshing state... (ID: 8u42yk)
aws_api_gateway_method.test_get: Refreshing state... (ID: agm-wosvgb26dh-8u42yk-GET)
aws_api_gateway_integration.test: Refreshing state... (ID: agi-wosvgb26dh-8u42yk-GET)
aws_api_gateway_deployment.test: Refreshing state... (ID: qhgl0a)
aws_api_gateway_stage.time: Refreshing state... (ID: ags-wosvgb26dh-test)
aws_api_gateway_method_settings.test: Refreshing state... (ID: wosvgb26dh-test-*/*)
aws_api_gateway_stage.time: Modifying... (ID: ags-wosvgb26dh-test)
  tags.%:       "0" => "1"
  tags.project: "" => "ProjectName"
aws_api_gateway_stage.time: Modifications complete after 1s (ID: ags-wosvgb26dh-test)

Apply complete! Resources: 0 added, 1 changed, 0 destroyed.
```

#### At update
First apply after `aws_api_gateway_deployment` update lead to remove `aws_api_gateway_stage` and `aws_api_gateway_method_settings` configuration. 
Need to re-apply to re-create it.

```
$ terraform taint aws_api_gateway_deployment.test
The resource aws_api_gateway_deployment.test in the module root has been marked as tainted!
 
$ terraform apply -auto-approve
aws_api_gateway_rest_api.test: Refreshing state... (ID: wosvgb26dh)
aws_api_gateway_resource.test: Refreshing state... (ID: 8u42yk)
aws_api_gateway_method.test_get: Refreshing state... (ID: agm-wosvgb26dh-8u42yk-GET)
aws_api_gateway_deployment.test: Refreshing state... (ID: qhgl0a)
aws_api_gateway_integration.test: Refreshing state... (ID: agi-wosvgb26dh-8u42yk-GET)
aws_api_gateway_method_settings.test: Refreshing state... (ID: wosvgb26dh-test-*/*)
aws_api_gateway_stage.time: Refreshing state... (ID: ags-wosvgb26dh-test)
aws_api_gateway_deployment.test: Destroying... (ID: qhgl0a)
aws_api_gateway_deployment.test: Destruction complete after 0s
aws_api_gateway_deployment.test: Creating...
  created_date:  "" => "<computed>"
  execution_arn: "" => "<computed>"
  invoke_url:    "" => "<computed>"
  rest_api_id:   "" => "wosvgb26dh"
  stage_name:    "" => "test"
aws_api_gateway_deployment.test: Creation complete after 1s (ID: v7ue28)
aws_api_gateway_stage.time: Modifying... (ID: ags-wosvgb26dh-test)
  deployment_id: "qhgl0a" => "v7ue28"
aws_api_gateway_stage.time: Modifications complete after 0s (ID: ags-wosvgb26dh-test)

Apply complete! Resources: 1 added, 1 changed, 1 destroyed.
```
Add this point, the tag in the stage on AWS console is removed
And a re-apply make updates on "aws_api_gateway_stage"
```
terraform apply -auto-approve
aws_api_gateway_rest_api.test: Refreshing state... (ID: wosvgb26dh)
aws_api_gateway_resource.test: Refreshing state... (ID: 8u42yk)
aws_api_gateway_method.test_get: Refreshing state... (ID: agm-wosvgb26dh-8u42yk-GET)
aws_api_gateway_integration.test: Refreshing state... (ID: agi-wosvgb26dh-8u42yk-GET)
aws_api_gateway_deployment.test: Refreshing state... (ID: v7ue28)
aws_api_gateway_stage.time: Refreshing state... (ID: ags-wosvgb26dh-test)
aws_api_gateway_method_settings.test: Refreshing state... (ID: wosvgb26dh-test-*/*)
aws_api_gateway_stage.time: Modifying... (ID: ags-wosvgb26dh-test)
  tags.%:       "0" => "1"
  tags.project: "" => "ProjectName"
aws_api_gateway_method_settings.test: Creating...
  method_path:                      "" => "*/*"
  rest_api_id:                      "" => "wosvgb26dh"
  settings.#:                       "" => "1"
  settings.0.throttling_rate_limit: "" => "1000"
  stage_name:                       "" => "test"
aws_api_gateway_method_settings.test: Creation complete after 0s (ID: wosvgb26dh-test-*/*)
aws_api_gateway_stage.time: Modifications complete after 0s (ID: ags-wosvgb26dh-test)

Apply complete! Resources: 1 added, 1 changed, 0 destroyed.
```

### Steps to Reproduce

1. `terraform apply`
2. `terraform import aws_api_gateway_stage.time [id]/test`
3. `terraform apply`
4. `terraform taint aws_api_gateway_deployment.test`
5. `terraform apply`
6. `terraform apply`

**Comment by fsoikin** [comment]
Is there a workaround currently?

**Comment by jurajseffer** [comment]
Not sure if it makes a difference here but deployment should depend on `aws_api_gateway_integration` not method.

**Comment by jurajseffer** [comment]
It looks like `aws_api_gateway_deployment` should interpolate `aws_api_gateway_stage.stage_name` because deployment auto creates a stage to which it should be deploying if it doesn't exist. Create stage, first, then deployment.

**Comment by mattbarry** [comment]
@jurajseffer how would you set `xray_tracing_enabled = true` on a stage if you use deployment to create it?

**Comment by jurajseffer** [comment]
@mattbarry I guess you can't without importing it. It seems to me that this is an AWS limitation - the logic of their resources is weird - stage requires a deployment and deployment requires a stage or auto creates it.

**Comment by mattbarry** [comment]
@jurajseffer for what it's worth, I did (somehow) get it working yesterday. AWS seems to be able to make sense of this:
```
resource "aws_api_gateway_stage" "stage" {
  stage_name    = "${aws_api_gateway_deployment.deployment.stage_name}"
  rest_api_id   = "${aws_api_gateway_rest_api.api.id}"
  deployment_id = "${aws_api_gateway_deployment.deployment.id}"
  xray_tracing_enabled = true
}

resource "aws_api_gateway_deployment" "deployment" {
  rest_api_id = "${aws_api_gateway_rest_api.api.id}"
  stage_name  = "${local.stage_name}"
}

resource "aws_api_gateway_method_settings" "settings" {
  rest_api_id = "${aws_api_gateway_rest_api.api.id}"
  stage_name  = "${aws_api_gateway_deployment.deployment.stage_name}"
  method_path = "*/*"
  depends_on = ["aws_api_gateway_stage.stage", "aws_api_gateway_deployment.deployment"]
}
```

**Comment by jurajseffer** [comment]
@mattbarry that's interesting because I believe I had a conflict of "stage already exists" when both deployment and stage were used.

**Comment by mattbarry** [comment]
@jurajseffer I did too, but this seems to be working. When I get some time I'll try to figure out why.

**Comment by hs-rendov** [comment]
guys, any news about this issue - seems there's no any workaround for now. empty stage_name isn't working using v2.23.0 -- getting "Error creating API Gateway Stage: BadRequestException: Stage name must be non-empty" or "Error creating API Gateway Stage: ConflictException: Stage already exists" if it is set.

**Comment by evaryont** [comment]
This is still happening, and is exacerbated when trying to fix #162 (for now, I'm using `timestamp()` in the description) -- now every apply is creating a new stage and thus losing the settings. I could run it again...but then that would create *another* new stage.

**Comment by KenBerg75** [comment]
The way we have been handling this is by using an origin stage for the deployment which is only used to stage the deployment object. We then create the actual stages we would use based on that deployment. 

```hcl

# The un-used stage simply exists to create the deployment
resource "aws_api_gateway_deployment" "rest_api_deployment" {
  rest_api_id       = aws_api_gateway_rest_api.rest_api_definition.id
  stage_name        = "origin"
  stage_description = "This stage is used to mount the deployment before rolling it to the active stages"

  lifecycle {
    create_before_destroy = true
  }
}

resource "aws_api_gateway_stage" "rest_api_dev_stage" {
  stage_name    = "DEV"
  rest_api_id   = aws_api_gateway_rest_api.rest_api_definition.id
  deployment_id = aws_api_gateway_deployment.rest_api_deployment.id
  xray_tracing_enabled = true

  variables = {
       ...
  }
}

resource "aws_api_gateway_stage" "rest_api_prd_stage" {
  stage_name    = "PRD"
  rest_api_id   = aws_api_gateway_rest_api.rest_api_definition.id
  deployment_id = aws_api_gateway_deployment.rest_api_deployment.id
  xray_tracing_enabled = true

  variables = {
       ...
  }
}

```

Not sure if thats the right way, but it works for us.

**Comment by bflad** [comment]
Hi folks 👋  Quite a lot has changed since this issue was initially filed, but all positive news.

Version 2.61.0 of the Terraform AWS Provider awhile ago made the `aws_api_gateway_deployment` resource `stage_name` argument optional, which solves both the original problems of this issue. When that value is configured, creating a deployment also creates the stage (why there cannot be an `aws_api_gateway_stage` with the same name). When the deployment resource needed to recreate (for a redeployment), it would delete and recreate the stage (why the stage and method settings disappeared unexpectedly). Both of these behaviors are confusing and we would recommend not using the `stage_name` argument, instead using just the `aws_api_gateway_stage` resource. To prevent other redeployment issues, also note that we recommend enabling `lifecycle` `create_before_destroy` as well with the `aws_api_gateway_deployment` resource.

To summarize these in a quick configuration example:

```terraform
resource "aws_api_gateway_deployment" "example" {
  rest_api_id = aws_api_gateway_rest_api.example.id

  triggers = {
    redeployment = # ... references to upstream resources (see issue link below) ...
  }

  lifecycle {
    create_before_destroy = true
  }
}

resource "aws_api_gateway_stage" "example" {
  deployment_id = aws_api_gateway_deployment.example.id
  rest_api_id   = aws_api_gateway_rest_api.example.id
  stage_name    = "example"
}
```

Please see https://github.com/hashicorp/terraform-provider-aws/issues/11344#issuecomment-765138213 for a lengthy writeup about these recommendations.
